crates/cubecl-runtime/src/memory_management/base.rs

```diff
@@ -1,5 +1,56 @@
+use core::fmt::Debug;
+
 use crate::storage::{ComputeStorage, StorageHandle, StorageId};
 
+/// Amount of memory in use by this allocator
+/// and statistics on how much memory is reserved and
+/// wasted in total.
+pub struct MemoryUsage {
+    /// The number of allocations currently active.
+    pub number_allocs: usize,
+    /// The number of bytes that are currently actually in use.
+    ///
+    /// This doesn't include any padding or other memory that needs to be
+    /// reserved, and is the minimum amount of memory that could possible
+    /// be allocated.
+    pub bytes_in_use: usize,
+    /// The amount of bytes used for padding memory in currently active allocations.
+    pub bytes_padding: usize,
+    /// The total amount of memory reserved on the device.
+    ///
+    /// This will be at least as much as bytes_in_use but in practice will
+    /// be higher, as allocations reserve memory for future allocations
+    /// and for padding.
+    pub bytes_reserved: usize,
+}
+
+impl MemoryUsage {
+    pub(crate) fn combine(&self, other: MemoryUsage) -> MemoryUsage {
+        MemoryUsage {
+            number_allocs: self.number_allocs + other.number_allocs,
+            bytes_in_use: self.bytes_in_use + other.bytes_in_use,
+            bytes_padding: self.bytes_padding + other.bytes_padding,
+            bytes_reserved: self.bytes_reserved + other.bytes_reserved,
+        }
+    }
+}
+
+impl std::fmt::Display for MemoryUsage {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        // In the future it'd be nice if MemoryUsage also held some stats about say,
+        // the 5 biggest allocations, to show when you an OOM.
+        let usage_percentage = (self.bytes_in_use as f32 / self.bytes_reserved as f32) * 100.0;
+        let padding_percentage = (self.bytes_padding as f32 / self.bytes_in_use as f32) * 100.0;
+        writeln!(f, "Memory Usage Report:")?;
+        writeln!(f, "  Number of allocations: {}", self.number_allocs)?;
+        writeln!(f, "  Bytes in use: {} bytes", self.bytes_in_use)?;
+        writeln!(f, "  Bytes used for padding: {} bytes", self.bytes_padding)?;
+        writeln!(f, "  Total bytes reserved: {} bytes", self.bytes_reserved)?;
+        writeln!(f, "  Usage efficiency: {:.2}%", usage_percentage)?;
+        writeln!(f, "  Padding overhead: {:.2}%", padding_percentage)
```

Comment:
Could we pretty format the bytes (kb, Mb, Gb, etc.)? It would be easier to debug.

Reply:
Yeah good idea! Added some formatting. Went with base 1000 but some other tools (nvidia-smi) use kibi / mebibytes, not kilo/megabytes. Idk what's easier for people

Reply:
Good enough I guess